File: nanopay/src/net/nanopay/fx/FXService.java
package net.nanopay.fx;

import foam.core.ContextAwareSupport;
import foam.dao.DAO;
import foam.nanos.NanoService;

public class FXService
    extends ContextAwareSupport
    implements FXServiceInterface, NanoService {

  private final FXServiceProvider fxServiceProvider;
  protected DAO fxQuoteDAO_;

  public FXService(final FXServiceProvider fxServiceProvider) {
    this.fxServiceProvider = fxServiceProvider;
  }

  public FXQuote getFXRate(String sourceCurrency, String targetCurrency,
      double sourceAmount, String direction, String valueDate, long user) throws RuntimeException {
    FXQuote fxQuote = this.fxServiceProvider.getFXRate(sourceCurrency, targetCurrency, sourceAmount, direction, valueDate, user);
    if ( null != fxQuote ) {
       fxQuote = (FXQuote) fxQuoteDAO_.put_(getX(), fxQuote);
    }

    return fxQuote;
  }

  public Boolean acceptFXRate(String quoteId, long user) throws RuntimeException {
    FXQuote quote = (FXQuote) fxQuoteDAO_.find(Long.parseLong(quoteId));
    if  ( null != quote ) {
      Boolean accepted = this.fxServiceProvider.acceptFXRate(quote.getExternalId(), user);
      if ( accepted ) {
        quote.setStatus(ExchangeRateStatus.ACCEPTED.getName());
        return true;
      }
    }

    return false;
  }

  public void start() {
    fxQuoteDAO_ = (DAO) getX().get("fxQuoteDAO");
  }

}

File: nanopay/src/net/nanopay/fx/FXService.js
/**
 * NANOPAY CONFIDENTIAL
 *
 * [2020] nanopay Corporation
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of nanopay Corporation.
 * The intellectual and technical concepts contained
 * herein are proprietary to nanopay Corporation
 * and may be covered by Canadian and Foreign Patents, patents
 * in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from nanopay Corporation.
 */

foam.INTERFACE({
  package: 'net.nanopay.fx',
  name: 'FXService',
  methods:  [
    {
      name: 'getFXRate',
      type: 'net.nanopay.fx.FXQuote',
      async: true,
      javaThrows: ['java.lang.RuntimeException'],
      args: [
        {
          name: 'sourceCurrency',
          type: 'String'
        },
        {
          name: 'targetCurrency',
          type: 'String'
        },
        {
          name: 'sourceAmount',
          type: 'Long'
        },
        {
          name: 'destinationAmount',
          type: 'Long'
        },
        {
          type: 'String',
          name: 'fxDirection',
        },
        {
          name: 'valueDate',
          type: 'String'// TODO: investigate why java.util.dat can't be used here
        },
        {
          type: 'Long',
          name: 'user'
        },
        {
          type: 'String',
          name: 'fxProvider'
        }
      ]
    },
    {
      name: 'getFXSpotRate',
      type: 'Double',
      async: true,
      javaThrows: ['java.lang.RuntimeException'],
      args: [
        {
          name: 'sourceCurrency',
          type: 'String'
        },
        {
          name: 'targetCurrency',
          type: 'String'
        },
        {
          type: 'Long',
          name: 'user'
        }
      ]
    },
    {
      name: 'acceptFXRate',
      type: 'Boolean',
      async: true,
      javaThrows: ['java.lang.RuntimeException'],
      args: [
        {
          name: 'quoteId',
          type: 'String'
        },
        {
          type: 'Long',
          name: 'user'
        }
      ]
    }
  ]
});

foam.CLASS({
  package: 'net.nanopay.fx',
  name: 'ProxyFXService',
  implements: [
    'net.nanopay.fx.FXService'
  ],
  properties: [
    {
      class: 'Proxy',
      of: 'net.nanopay.fx.FXService',
      name: 'delegate'
    }
  ]
});
